compiler/rustc_expand/src/proc_macro.rs

```diff
@@ -181,3 +198,102 @@ impl MultiItemModifier for DeriveProcMacro {
         ExpandResult::Ready(items)
     }
 }
+
+pub(super) fn provide_derive_macro_expansion<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: (LocalExpnId, Svh, &'tcx TokenStream),
+) -> Result<&'tcx TokenStream, ()> {
+    let (invoc_id, _macro_crate_hash, input) = key;
+
+    with_context(|(ecx, client)| expand_derive_macro(tcx, invoc_id, input, ecx, *client))
+}
+
+type CLIENT = pm::bridge::client::Client<pm::TokenStream, pm::TokenStream>;
+
+fn expand_derive_macro<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    invoc_id: LocalExpnId,
+    input: &'tcx TokenStream,
+    ecx: &mut ExtCtxt<'_>,
+    client: CLIENT,
+) -> Result<&'tcx TokenStream, ()> {
+    let invoc_expn_data = invoc_id.expn_data();
+    let span = invoc_expn_data.call_site;
+    let event_arg = invoc_expn_data.kind.descr();
+    let _timer = tcx.sess.prof.generic_activity_with_arg_recorder(
+        "expand_derive_proc_macro_inner",
+        |recorder| {
+            recorder.record_arg_with_span(tcx.sess.source_map(), event_arg.clone(), span);
+        },
+    );
+
+    let proc_macro_backtrace = ecx.ecfg.proc_macro_backtrace;
+    let strategy = crate::proc_macro::exec_strategy(tcx.sess);
+    let server = crate::proc_macro_server::Rustc::new(ecx);
+
+    match client.run(&strategy, server, input.clone(), proc_macro_backtrace) {
+        Ok(stream) => Ok(tcx.arena.alloc(stream) as &TokenStream),
+        Err(e) => {
+            tcx.dcx().emit_err({
+                errors::ProcMacroDerivePanicked {
+                    span,
+                    message: e.as_str().map(|message| errors::ProcMacroDerivePanickedHelp {
+                        message: message.into(),
+                    }),
+                }
+            });
+            Err(())
+        }
+    }
+}
+
+// based on rust/compiler/rustc_middle/src/ty/context/tls.rs
+thread_local! {
+    /// A thread local variable that stores a pointer to the current `CONTEXT`.
+    static TLV: Cell<(*mut (), Option<CLIENT>)> = const { Cell::new((ptr::null_mut(), None)) };
```

Comment:
This looks more like `scoped_thread_local!`, because we put a value into it outside the query call, and then retrieve it immediately inside the call and don't need it anymore.
Maybe it will help to avoid some manual type erasure.